### DISCUSSION TITLE
[@types/ssh2] Issue with readdir callback retuning FileEntry

### DISCUSSION THREAD

**Discussion by gdelory** [DefinitelyTyped/DefinitelyTyped#67360] [Issues with a @types package]
Hi,

I feel like there is something wrong with s[ftp.readir callback type](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/451dc8fc19383bc12af59522020e571957f1684e/types/ssh2/index.d.ts#L1547). We are in the process of migrating to TypeScripr (from pure Javascript), and the code we had using the `attrs.isDirectory()` and `attrs.isSymbolicLink()` and is not compiling, while it works in Javascript.

Here the code failing (ssh2-issue.ts):

```
import { Client } from 'ssh2'

const main = async () => {
  const conn = new Client()
  conn.on('ready', () => {
    conn.sftp(async (err, sftp) => {
      if (err) {
        console.error(err)
      } else {
        sftp.readdir('/home/g/d/gdelory', (err, res) => {
          for (let i = 0; i < res.length; i++) {
            const e = res[i]
            console.log(e.attrs.isDirectory())
            console.log(e.attrs.isSymbolicLink())
          }
          conn.end()
        })
      }
    })
  }).on('error', console.error).connect({
    host: 'rtpgsa.ibm.com',
    port: 22,
    username: 'gdelory',
    password: '*****************'
  })
}
main()
```

With the following errors:
```
src/ssh2-issue.ts(13,33): error TS2339: Property 'isDirectory' does not exist on type 'Attributes'.
src/ssh2-issue.ts(14,33): error TS2339: Property 'isSymbolicLink' does not exist on type 'Attributes'.
```

Note that the exact same code works fine in JS, and if I run this following snippet:

```
import { Client } from 'ssh2'

const main = async () => {
  const conn = new Client()
  conn.on('ready', () => {
    conn.sftp(async (err, sftp) => {
      if (err) {
        console.error(err)
      } else {
        sftp.readdir('/home/g/d/gdelory', (err, res) => {
          for (let i = 0; i < res.length; i++) {
            const e = res[i]
            console.log(e.attrs)
          }
          conn.end()
        })
      }
    })
  }).on('error', console.error).connect({
    host: 'rtpgsa.ibm.com',
    port: 22,
    username: 'gdelory',
    password: '***************'
  })
}
main()
```
It outputs the following:

```
Stats {
  mode: 33571312,
  uid: 544506,
  gid: 988871,
  size: 4096,
  atime: 1699529953,
  mtime: 1673417659,
  extended: undefined
}
Stats {
  mode: 33572288,
  uid: 544506,
  gid: 988871,
  size: 4096,
  atime: 1699529953,
  mtime: 1646028984,
  extended: undefined
}
Stats {
  mode: 33572349,
  uid: 544506,
  gid: 988871,
  size: 4096,
  atime: 1699529953,
  mtime: 1646028986,
  extended: undefined
}
```

Proving that `attrs` is of type [Stats](https://github.com/mscdex/ssh2/blob/master/SFTP.md#stats), not [Attrs](https://github.com/mscdex/ssh2/blob/master/SFTP.md#attrs).

Also the code I found in the rpo seems to say the same. [Here](https://github.com/mscdex/ssh2/blob/483c59955e2d9ab65ed2543077b0df79ef270260/lib/protocol/SFTP.js#L2589) an instance of Stats is created, in the [readAttrs](https://github.com/mscdex/ssh2/blob/483c59955e2d9ab65ed2543077b0df79ef270260/lib/protocol/SFTP.js#L2570) function which seems to be called [here](https://github.com/mscdex/ssh2/blob/483c59955e2d9ab65ed2543077b0df79ef270260/lib/protocol/SFTP.js#L2962) when using readdir.
 
So I think [these lines](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/451dc8fc19383bc12af59522020e571957f1684e/types/ssh2/index.d.ts#L1360-L1364):
```
export interface FileEntry {
    filename: string;
    longname: string;
    attrs: Attributes;
}
```
Should be
```
export interface FileEntry {
    filename: string;
    longname: string;
    attrs: Stats;
}
```
What you do think?

Version info:

I'm on the last version of ssh2 (1.1.14):
```
~/wfm$ npm list ssh2
wfm@2.0.0 /home/gui/wfm
└── ssh2@1.14.0
```
and I think the last version the type module:
```
~/wfm$ npm list @types/ssh2
wfm@2.0.0 /home/gui/wfm
└── @types/ssh2@1.11.16
```

[a comment by an automated account is not shown]

**Comment by dhensby** [comment]
Seems like a valid change. 👍

**Comment by gdelory** [comment]
Do you want me to try to submit a PR? I just wasn't sure it's true for every method returning/using `FileEntry`. Although looking at the code, since [readAttrs](https://github.com/mscdex/ssh2/blob/483c59955e2d9ab65ed2543077b0df79ef270260/lib/protocol/SFTP.js#L2570) seems to be used every time `attrs` are created, I think it should be?

**Comment by gdelory** [comment]
Created issue https://github.com/DefinitelyTyped/DefinitelyTyped/issues/67521 and PR https://github.com/DefinitelyTyped/DefinitelyTyped/pull/67522

**Comment by slifty** [comment]
I know the fix to this concern is now merged but wanted to raise an issue the solution created for a project I'm a part of: https://github.com/PermanentOrg/sftp-service/pull/330#issuecomment-1839431140

In order to create a FileEntry that is being passed to `name`, one now needs to manually re-implement (or mock) the functionality of those `Stats` methods, since the class itself is not exposed by the ssh2 library and the types now require implementation of those methods.

Would it be reasonable to create two FileEntry types, one with Stats (where the FileEntry is being returned by the ssh2 library, and has the Stats functions populated), and one without Stats (e.g. where the functions are not necessarily present).

This would mean that the signature of, for instance `readdir`'s callback could be coded as having `Stats` but the instance of `sftpConnection.name` would *NOT* require stats (and indeed only requires `Attrs`)

**Comment by dhensby** [comment]
@slifty I'm a little confused about what you're asking. How can we have two types/interfaces that are named the same but are not the same?

I see the `Stats` class is not exported, but it's not a public interface for one to be using/constructing themselves, so I don't see that is a problem (or am I mistaken?).

It seems like your issue is a typing one, and not really an implementation one. You don't want to have to mock out the additional functions on the attributes property, even though the typings do reflect reality. If you just want to satisfy they type engine you can just do `attributes as unknown as Stats`...

**Reply by slifty** [reply]
I believe that solution would bypass type safety so it's not an ideal one for us (even though I agree it would compile).

Sorry for not explaining clearly what I'm talking about -- my understanding is that ssh2 has two types of FileEntity, in practice.

1. The `FileEntity` that is being *returned* by ssh2 has `attrs` of type `Stats` (which extends `Attributes`)
2. The `FileEntity` that is being *consumed* by ssh2 does NOT require `Stats` but merely requires `Attributes`.

The PR that has been merged is overconstraining around the `FileEntities` that are being *consumed* -- it is requiring those FileEntities to be `Stats` even though ssh2 does not actually use / expect / require any of the helper functions.

What I'm proposing is that there be two FileEntity types -- one for the values being returned, one for the values being consumed.

Here's what I'm talking about:

```
export interface FileEntry {
    filename: string;
    longname: string;
    attrs: Attributes;
}

export interface FileEntryWithStats extends FileEntry {
    attrs: Stats;
}

// The signature of the `readdir` callback would properly signal that the file entry has `Stats` and so 
// the original issue's problem would be addressed
readdir(location: string | Buffer, callback: (err: Error | undefined, list: FileEntryWithStats[]) => void): void;

// The signature of the `name` however would just be a normal FileEntry, and therefore would NOT require
// anybody to bypass type safety OR mock unnecessary functions when invoking `name()`
// This is more correct since the implementation of `name` does not actually require a FileEntry with Stats,
// but rather only requires a FileEntry with Attributes.
name(reqId: number, names: FileEntry[]): void;
```

Does that make sense?

**Reply by dhensby** [reply]
ok - cool, that makes sense - do you want to open a PR for it?

**Reply by slifty** [reply]
Happy to do so!

**Reply by gdelory** [reply]
Sorry about, I didn't know FileEntry were also used as input parameters (we don't use those functions). If what's expected as input is different from where being returned, it completely makes sense to have two distinct interfaces. Thanks for the PR and sorry for overlooking that.

**Reply by slifty** [reply]
@gdelory oh no problem at all (and sorry I was so slow to actually open this)

PR for the proposed additional improvement is here: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/68263
